### ISSUE TITLE
Insufficient Hecke eigenvalues

### ISSUE THREAD

**Issue by jvoight** [LMFDB/lmfdb#1656]
@danyasaki reported that certain HMFs do not have as many Hecke eigenvalues as prime ideals; he found the following:

2.2.85.1-180.2-m
2.2.85.1-180.2-n
2.2.85.1-180.2-o
2.2.85.1-180.2-p
2.2.85.1-180.2-q
2.2.85.1-180.3-m
2.2.85.1-180.3-n
2.2.85.1-180.3-o
2.2.85.1-180.3-p
2.2.85.1-180.3-q

(There does not appear to be any problem with forms with level 180.1.)

**Comment by jvoight** [comment]
@AurelPage, @JohnCremona I do not have these ideals in my text files--the largest ideal NN consider has norm 180 and has NN cap ZZ = (30).  Were these added by one of you, or am I not remembering something I did?

**Comment by AurelPage** [comment]
@jvoight Are they Galois conjugate of ideals that were present in your text files? If yes, it is possible that we added them when we added Galois conjugate forms (I cannot check right now but I will).

**Comment by jvoight** [comment]
The ideals 180.2 and 180.3 are Galois conjugate, but they are not conjugate to forms in my data files.  So something was added beyond these (possibly by me).

**Comment by JohnCremona** [comment]
@jvoight I think these must have been added by you since all Aurel and I added were Galois conjugates of forms you had uploaded.

**Comment by jvoight** [comment]
OK, then lord knows what I did.  

It is easy enough to fix just these, but I would rather double check all HMFs to make sure that there are as many Hecke eigenvalues as primes.  I don't see it causing a problem with the LMFDB website, and actually maybe we'll want to allow some forms to have fewer eigenvalues than primes in the future.  

So I'll put this on my todo list, but it's low priority.

**Comment by AndrewVSutherland** [comment]
@jvoight I just want to be sure that I understand -- are yousaying that this problem has not impact on what the user sees on the website and, in particular, it is not causing the LMFDB to display information that is mathematically incorrect?  If so then I agree it is low priority and we can defer this to Release 1.2 or 2.0.

But if there is false information being displayed (or available for download) than I think we should remove the affected records from the database now (or modify the code to hide them, but I think that is a less pleasant solution -- it has caused us no end of grief in the classical modular forms code).

**Comment by JohnCremona** [comment]
Despite the possible non-urgency, it will be easy to check this systematically since both the primes in the fields collection and the Hecke eigenvalues in the forms collection are lists of strings, so comaring the lengths is easy.
I just tried this with 2.2.5.1 where tere are 1233 primes in the list, but *very* many newforms have fewer hecke eigenvalues.
So unless I am misunderstanding something,  this is not a bug.  It would only be a bug if any of the code *expected* there to be as many eigenvalues as primes in the field collection.

**Comment by AndrewVSutherland** [comment]
If it just a question of completeness of the data, then I agree that it is not urgent, but perhaps there should be an extent knowl (linked to in  the learnmore box) that explains that the lists of eigenvalues are incomplete?

**Comment by JohnCremona** [comment]
But no individual HMF makes any claim about how many eigenvalues it has stored (e.g. a random one: http://localhost:37777/ModularForm/GL2/TotallyReal/2.2.33.1/holomorphic/2.2.33.1-199.2-a).  Admittedly, they probably should since we have a box inviting users to ask for more eigenvalues, without giving any idea of how many are available.  For the link above I just typed 1000000 into the box and there were more displayed (primes of norm up to 10000 by the looks of it), but not a million, and nothing went wrong (the code has the expression eigs = eigs[:min(len(eigs), numeigs)] at some relevant point).

**Comment by AndrewVSutherland** [comment]
@JohnCremona I don't think its necessary to say anything specific about how many eigenvalues are stored for each form, but as a user of the site, if I see a page with a list of eigenvalues ordered by the norm of the prime I would not expect there to be gaps, or eignevalues for only some of the primes of a given norm.  If I did happen to notice an eigenvalue missing within the range of eigenvalues listed and clicked on the "Completeness of this data" link (e.g. on http://www.lmfdb.org/ModularForm/GL2/TotallyReal/2.2.33.1/holomorphic/2.2.33.1-199.2-a), I would expect to see at least some mention of the fact that there could be gaps (if I did not I would probably submit a bug report, just as Dan Yasaki did).

**Comment by JohnCremona** [comment]
OK, so the consistency we seek is not that all newforms have the same number of eigenvalues but that the primes for which we do have eigenvalues form a set of the form {P: P.norm()<B} for some B.  I would certainly expect that to be true of the the data we have but I *know* that it is not since Aurel and I discovered when adding Galois conjugates that the set of primes for which we have eigenvalues is not always closed under conjugation!  This could well be exactly the issue which Dan noticed.

The simplest way to achieve this without doing more computations would be to actually delete stored eigenvalues for any prime for which we do not have eigenvalues for all primes of that norm.  At the same time as doing that we could store for each newform the maximal norm of primes for which we have eigenvalues (OK, not really necessary, since we can always extract that from looking at the number of eigenvalues we have and then looking at the prime of that index), and display that together with the number of eigenvalues on the newform's web page.

I think I prefer this option to some undetermined wait for more data to be computed.  For the future, the upload script could check that the eigenvalues being uploaded have the no-gaps property.

[I am definitely assuming that if we have an a_P for some prime P then we have all a_Q for primes Q of norm strictly less than P!]

**Comment by AndrewVSutherland** [comment]
@JohnCremona +1 for avoiding "some undetermined wait".

**Comment by JohnCremona** [comment]
@AurelPage Would you be willing to do what I suggest above, i.e. by deleting some entries from the end of the heck_eigenvalues list for each newform where necessary, get to the state where the primes for which we do have eigenvalues stored are consistent in the sense that if we have any a_P for any P of a given norm then we have all a_Q for all Q of norm <= P.norm()?   I thought it might be rather similar to something else you did with respect to conjugation.

Once that is done we can add to the newforms display the information about how many primes and the max norm for which we have eigenvalues.

**Comment by AndrewVSutherland** [comment]
I've tagged this for Release 1.1.

**Comment by jvoight** [comment]
Sorry for the "undetermined wait" for this issue!  Let's kill this off at Warwick next month.

**Comment by jvoight** [comment]
The issue has been resolved, as follows.  

In my HMF upload script, I cropped the list of Hecke eigenvalues to 2 MB because at the time, this was the limit for the list of strings.  There are other ways now if we wanted to store all of this data (chunked files), but it's not clear we do: these correspond to huge Hecke fields and enormous eigenvalues which cannot be displayed anyway.  

Aurel is now going through to ensure that whenever there is an a_pp there are all a_qq with N(pp) = N(qq), in other words, we have all of a given prime norm.  

I'll add a comment now in the data completeness knowl to this effect, but we should be able to close this issue soon.
